src/read.rs

```diff
@@ -1684,18 +1677,221 @@ pub fn read_zipfile_from_stream<'a, R: Read>(reader: &'a mut R) -> ZipResult<Opt
+        ReadBufferThenForward {
+            reader,
+            buffer,
+            index: 0,
+        }
+    }
+}
+
+impl<'a, T: Read> Read for ReadBufferThenForward<'a, T> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        let buffer_start = std::cmp::min(self.index, buf.len());
+        let data_left_in_buffer = buf.len() - buffer_start;
+        let buffer_end = std::cmp::min(self.buffer.len(), buffer_start + data_left_in_buffer);
+
+        let next_buffer_data = &self.buffer[buffer_start..buffer_end];
+        let count_copy_from_buffer = std::cmp::min(next_buffer_data.len(), buf.len());
+
+        buf[0..count_copy_from_buffer]
+            .copy_from_slice(&next_buffer_data[0..count_copy_from_buffer]);
+
+        self.index += count_copy_from_buffer;
+        if self.index > buf.len() && self.index > 0 {
+            // empty buffer
+            self.index = 0;
+            self.buffer.clear();
+            self.buffer.shrink_to_fit();
+        }
+
+        let remaining_bytes = buf.len() - count_copy_from_buffer;
+        let result = if remaining_bytes > 0 {
+            self.reader.read(&mut buf[count_copy_from_buffer..])?
+        } else {
+            0
+        };
+
+        Ok(count_copy_from_buffer + result)
+    }
+}
+
+struct ReadTillDataDescriptor<T: Read> {
+    reader: T,
+    data_descriptor_found: Option<ZipDataDescriptor>,
+    look_ahead_buffer: Vec<u8>,
+    number_read_total_actual: usize, // without look ahead buffer
+}
+
+impl<T: Read> ReadTillDataDescriptor<T> {
+    pub fn new(mut reader: T) -> ZipResult<Self> {
+        let mut look_ahead_buffer = [0u8; std::mem::size_of::<ZipDataDescriptor>()];
+
+        // fill the look ahead buffer
+        reader.read_exact(&mut look_ahead_buffer)?;
+
+        let look_ahead_buffer = Vec::from(look_ahead_buffer);
+
+        Ok(Self {
+            reader,
+            look_ahead_buffer,
+            data_descriptor_found: None,
+            number_read_total_actual: 0,
+        })
+    }
+}
+
+impl<T: Read> ReadTillDataDescriptor<T> {
+    fn has_found_data_descriptor(&self) -> Option<ZipDataDescriptor> {
+        if spec::Magic::from_first_le_bytes(&self.look_ahead_buffer)
+            == spec::Magic::DATA_DESCRIPTOR_SIGNATURE
+        {
+            // potentially found a data descriptor
+            // check if size matches
+            let data_descriptor = match ZipDataDescriptor::interpret(&self.look_ahead_buffer) {
+                Ok(data_descriptor) => data_descriptor,
+                Err(_) => return None,
+            };
+            let data_descriptor_size = data_descriptor.compressed_size;
+
+            if data_descriptor_size == self.number_read_total_actual as u32 {
+                return Some(data_descriptor);
+            }
+        }
+        None
     }
 
-    let limit_reader = (reader as &'a mut dyn Read).take(result.compressed_size);
+    fn read_a_byte(&mut self) -> io::Result<Option<u8>> {
```

Comment:
Reading one byte at a time like this won't be efficient. Instead, read a block into the look-ahead buffer and use `memchr::memmem::find` to check for a data descriptor signature. (Be sure to leave `size_of::<Magic>()` bytes in the buffer, in case the descriptor signature straddles two blocks!)